### PR TITLE
fix(emit): eliminate multiple cache files getting out of sync and fix racy emit

### LINKED ISSUE #13302: Emit, source map, and declaration cache in deno dir is racy across multiple processes
I've run into two caching issues while using deno the past two days where I was sure my code was ok, but I was getting type checking errors.

Upon closer inspection, our cache code is racy across multiple processes since the hash is stored in a separate file.

Also, looks like if you kill the process between these two statements then the emit won't be saved correctly and won't ever be reloaded until you change the source text:

https://github.com/denoland/deno/blob/12423e16b7df39b2bf995efd1376ba4b6ef41e02/cli/emit.rs#L458-L459

To solve this, we should store both the version hash along with each artifact in the same file. I've already started working on a fix.

### PR DESCRIPTION
**Pull request by dsherret**
1. Uses an sqlite database file to store the emit cache to eliminate multiple cache files getting out of sync.
2. The result of checking the emit is stored in memory and if valid is kept in memory then used in the rest of the emit. This prevents consulting the file system to check if the cache is valid and then reading from the file system later when the cache may no longer be valid, which was racy.

Supersedes #13303.

Closes #13302

WIP though basically done. I could do way more cleanup and have a better design, but wanted to make the minimal changes possible.

TODO:

- [ ] Ensure all calls to `get_emit_data` verify the source. Maybe an API change would be good to force this (ex. `ProcState`'s `SourceMapGetter` impl right now retrieves the source map from the cache, but doesn't do any source checks on it). I think though it probably needs more rework in that area and I will have to read the code more.
- [ ] coverage's `cover_files` does not verify the retrieved emit text is for the source

### CODE REVIEW

File: cli/emit_cache.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+  /// these stored here for mutating for testing purposes
+  emit_data_version: usize,
+  ts_build_info_version: usize,
+}
+
+impl SqliteEmitCache {
+  pub fn new(db_file_path: &Path) -> Result<Self, AnyError> {
+    let conn = Connection::open(db_file_path)?;
+    Self::from_connection(conn)
+  }
```
Comment (littledivy) [inline cli/emit_cache.rs:101]: Consider using these pragmas, they can significantly speed things up. https://github.com/denoland/deno/blob/56d0ca7f52d9bff7326bb2b9406ea1e0282a5fbe/ext/webstorage/lib.rs#L69-L81

Comment (dsherret) [reply cli/emit_cache.rs:101]: Thanks for letting me know! I added them 🙂

**Comment by jespertheend** [comment]
Any updates on this? I'm hoping this will fix https://github.com/denoland/deno/issues/12126

**Comment by KnorpelSenf** [comment]
Potentially fixes #14206, as well.

**Comment by dsherret** [comment]
Closing because this is really out of date. I added an `EmitCache` trait in #14925 (merged) and will follow up with another PR in a few weeks (before 1.24) that uses the sql cache from this PR (along with doing the same in /x/deno_cache).

**Pull request closed without merging**